File: src/libgql/src/libgql/parsers/file/server/ast.cpp
#include "./ast.hpp"

#include <optional>
#include <string>

namespace parsers::file::server::ast {
std::optional<DirectiveLocation> stringToDirectiveLocation(
    const std::string &str) {
    if (str == "SCHEMA")
        return DirectiveLocation::SCHEMA;
    else if (str == "SCALAR")
        return DirectiveLocation::SCALAR;
    else if (str == "OBJECT")
        return DirectiveLocation::OBJECT;
    else if (str == "FIELD_DEFINITION")
        return DirectiveLocation::FIELD_DEFINITION;
    else if (str == "ARGUMENT_DEFINITION")
        return DirectiveLocation::ARGUMENT_DEFINITION;
    else if (str == "INTERFACE")
        return DirectiveLocation::INTERFACE;
    else if (str == "UNION")
        return DirectiveLocation::UNION;
    else if (str == "ENUM")
        return DirectiveLocation::ENUM;
    else if (str == "ENUM_VALUE")
        return DirectiveLocation::ENUM_VALUE;
    else if (str == "INPUT_OBJECT")
        return DirectiveLocation::INPUT_OBJECT;
    else if (str == "INPUT_FIELD_DEFINITION")
        return DirectiveLocation::INPUT_FIELD_DEFINITION;
    return std::nullopt;
};
};  // namespace parsers::file::server::ast

File: rust/gql/libgql/src/parsers/schema/client/ast.rs
use std::{cell::RefCell, rc::Rc};

use crate::parsers::{
    file,
    schema::{server, shared},
};

#[derive(Debug)]
pub struct SpreadSelection {
    pub fragment: Rc<Fragment>,
}

#[derive(Debug)]
pub struct TypenameField {
    pub alias: Option<String>,
}

#[derive(Debug, derive_more::From)]
pub enum UnionSelection {
    TypenameField(TypenameField),
    SpreadSelection(SpreadSelection),
    ObjectConditionalSpreadSelection(ObjectConditionalSpreadSelection),
    UnionConditionalSpreadSelection(UnionConditionalSpreadSelection),
}

#[derive(Debug, derive_more::From)]
pub enum ObjectSelection {
    TypenameField(TypenameField),
    SpreadSelection(SpreadSelection),
    FieldSelection(FieldSelection),
}

#[derive(Debug)]
pub struct UnionFragmentSpec {
    pub r#type: Rc<RefCell<server::ast::Union>>,
    pub selections: Vec<UnionSelection>,
}

#[derive(Debug)]
pub struct ObjectFragmentSpec<T> {
    pub r#type: Rc<RefCell<T>>,
    pub selections: Vec<ObjectSelection>,
}

#[derive(Debug)]
pub struct ObjectConditionalSpreadSelection {
    pub r#type: Rc<RefCell<server::ast::ObjectType>>,
    pub selection: Rc<ObjectFragmentSpec<server::ast::ObjectType>>,
}

#[derive(Debug)]
pub struct UnionConditionalSpreadSelection {
    pub r#type: Rc<RefCell<server::ast::Union>>,
    pub selection: Rc<UnionFragmentSpec>,
}

#[derive(Debug, derive_more::From)]
pub enum FragmentSpec {
    Union(UnionFragmentSpec),
    Object(ObjectFragmentSpec<server::ast::ObjectType>),
    Interface(ObjectFragmentSpec<server::ast::Interface>),
}

#[derive(Debug)]
pub struct FieldSelection {
    pub name: String,
    pub alias: String,
    pub arguments:
        indexmap::IndexMap<String, shared::ast::FieldSelectionArgument>,
    pub selection: Option<Rc<FragmentSpec>>,
}

#[derive(Debug)]
pub struct Fragment {
    pub name: String,
    pub spec: FragmentSpec,
    pub source_text: String,
    pub hash: u64,
}

#[derive(Debug)]
pub struct Operation {
    pub r#type: file::client::ast::OpType,
    pub name: String,
    pub parameters: indexmap::IndexMap<
        String,
        shared::ast::FieldDefinition<shared::ast::InputFieldSpec>,
    >,
    pub fragment_spec: FragmentSpec,
    pub used_fragments: Vec<Rc<RefCell<Fragment>>>,
    pub source_text: String,
    pub parameters_hash: u64,
    pub fragment_spec_hash: u64,
}

#[derive(Debug)]
pub struct ClientDirective {
    pub name: String,
    pub arguments: indexmap::IndexMap<
        String,
        shared::ast::FieldDefinition<shared::ast::InputFieldSpec>,
    >,
    pub locations: Vec<file::client::ast::DirectiveLocation>,
}

#[derive(derive_more::From)]
pub enum ClientSchemaNode {
    Fragment(Rc<RefCell<Fragment>>),
    Operation(Rc<RefCell<Operation>>),
    ClientDirective(Rc<ClientDirective>),
}
